### PR TITLE
Variable step height + Degree reduction + Lookup refactor

### PR DESCRIPTION
**Pull request by Brechtpd**
Cleaned up version of experimental optimizations done here: https://github.com/Brechtpd/zkevm-circuits/pull/2.
Code depends on some small halo2 changes: https://github.com/appliedzkp/halo2/pull/54

- Extended domain now only 4x larger by limiting the max degree to 5.
- There are now only 49 lookups/row (down from 300+/row I think).
- Current parameters are chosen pretty much arbitrarily, with the circuit now being more than double as wide which I think will probably make sense (but remains to be seen). Best parameters need to be found by measuring against real blocks or minimizing prover time/gas.

Left for another PR:
- It's possible to allow certain cells to be used for multiple purposes (e.g. storage + byte lookup) like was done before, but could even be extended if it makes sense.
- There are a lot of byte lookups being done, but often that is just to load in words that are loaded from the stack/memory. Because these are rlc encoded and were already checked to contain only bytes it's probably possible not to recheck this when just loading them in again. Changing this would greatly help reduce the number of byte lookups.
- Check for complete expressions that are shared between opcodes and just share them directly in the circuit (add their state selectors together). Not sure how much that will end up saving though because most calculations in those cases should already be done once on the prover side (but at least needs an extra multiplication and addition for Horner's rule).
- Smarter cell assignment to maximize shared subexpressions between opcode implementations. Postpone assigning variables to cells until all opcode expressions are known so this can be optimized in code.

How much faster the circuit is now gets quite difficult because just using the number of rows doesn't really say much anymore, nevertheless using that metric the circuit seems to be 4x faster while for a lot of opcodes needing a lot less rows. We'll have to use something more meaningful to measure performance well (like number of ERC20 transfers/second, or number of uniswap trades/second, or worst case scenario gas/second).

I also made the circuit much wider so that opcodes require less rows. Opcodes that require a certain amount of resource (a certain lookup type, cell storage) much more than other opcodes still use a pretty high number of rows so that we don't need to make all other opcodes more expensive. But there is a practical balance to be found to do this well and so the configuration for this will need to be tweaked. `MAX_COPY_BYTES` was lowered to the same amount of bytes that already need to be copied in `CALLDATALOAD`/`MLOAD` so that the number of lookups required are the same. But with the copy circuit I guess this doesn't really matter anymore.
 
```
BeginTx: 10 rows
COPYCODETOMEMORY: 8 rows
COPYTOMEMORY: 8 rows
CopyToLogGadget: 8 rows
EndBlock: 2 rows
EndTx: 14 rows
ADD: 5 rows
BITWISE: 5 rows
BYTE: 3 rows
CALL: 13 rows
CALLVALUE: 2 rows
CALLDATACOPY: 3 rows
CALLDATALOAD: 8 rows
CALLDATASIZE: 2 rows
CALLER: 2 rows
CHAINID: 2 rows
CODECOPY: 3 rows
CMP: 5 rows
DUP: 2 rows
EXTCODEHASH: 2 rows
GAS: 2 rows
GASPRICE: 2 rows
ISZERO: 2 rows
JUMP: 2 rows
JUMPDEST: 2 rows
JUMPI: 2 rows
LOG: 2 rows
MEMORY: 5 rows
MSIZE: 2 rows
MUL_DIV_MOD: 8 rows
ORIGIN: 2 rows
PC: 2 rows
POP: 2 rows
PUSH: 9 rows
SELFBALANCE: 2 rows
SCMP: 5 rows
SIGNEXTEND: 4 rows
SLOAD: 2 rows
SSTORE: 2 rows
STOP: 2 rows
SWAP: 2 rows
BlockCTXU64: 2 rows
BlockCTXU160: 2 rows
BLOCKCTXU256: 2 rows
ErrorOutOfGasStaticMemoryExpansion: 3 rows
```

### CODE REVIEW

**Review by CPerezz — COMMENTED** [review]
Just an initial review with some comments and questions.

This looks awesome @Brechtpd :partying_face:

File: .github/proverCiScripts/execBench.sh
Comment (Brechtpd) [reply .github/proverCiScripts/execBench.sh:36]: Just making the stack size twice as large as the default as a temporary workaround: https://github.com/appliedzkp/zkevm-circuits/pull/417#issuecomment-1086465517

File: zkevm-circuits/src/evm_circuit.rs
Comment (Brechtpd) [reply zkevm-circuits/src/evm_circuit.rs:35]: Oh wasn't aware this was something that could be dangerous! I'm still pretty new to rust so don't always know what the best practices are. If it's something we want to avoid will certainly remove. Will also look into this some more so I better know the potential pitfalls. Thanks for the warning!

File: zkevm-circuits/src/evm_circuit.rs
Comment (Brechtpd) [reply zkevm-circuits/src/evm_circuit.rs:107]: That's probably be a good idea. I merged them into one to remove code duplication, but it does seem like it would be better to have two separate functions, with one only usable in tests, that call into this general one. The normal one always passing in `exact: false` and the test one always passing in `exact: true`. How does that sound?

File: zkevm-circuits/src/evm_circuit/util.rs
Comment (Brechtpd) [reply zkevm-circuits/src/evm_circuit/util.rs:120]: For sure could be sparse, but we only need to store up to 3 steps in memory (so not something that grows with the circuit size), so the amount of memory this takes is very small. This code is used in the actual prover to set the witness, so performance is important and I think using a simple array lookup is likely among the fastest solutions.

File: zkevm-circuits/src/evm_circuit/util.rs
Comment (Brechtpd) [reply zkevm-circuits/src/evm_circuit/util.rs:280]: So the main reason for this function to exist is to get some insights in what needs to be stored. Perhaps in the future something smarter can be done with this (like automatically find the optimal number of columns for each cell type), but for now it can be used to just print out the info to see what's going on and what is used. I'm using a `BTreeMap` here because otherwise if you run over the keys/values the order is always different and so it's harder to compare the results of different calls to this function without having to sort on the keys again. Hope this makes sense!

File: zkevm-circuits/src/evm_circuit/util.rs
Comment (Brechtpd) [reply zkevm-circuits/src/evm_circuit/util.rs:514]: Good point. `bytes` is a bad name here actually because it can be any value, will change the name to also make that more clear.

File: zkevm-circuits/src/evm_circuit/execution.rs
Comment (han0110) [reply zkevm-circuits/src/evm_circuit/execution.rs:239]: There are `COPYTOMEMORY` and `COPYCODETOMEMORY` still being tested with incomplete trace. In #419 I removed them because I believe they are covered by `CALLDATACOPY` and `CODECOPY`, and enable these checks.

I think it's okay to just keep them as TODO in this PR.

Comment (CPerezz) [reply zkevm-circuits/src/evm_circuit/execution.rs:239]: Ohhh I see.
Well, then let's file a `tech-debt` issue for this so that we don't forget no? 

WDYT?

File: .github/proverCiScripts/execBench.sh
Comment (CPerezz) [reply .github/proverCiScripts/execBench.sh:36]: Ohhh!! Right!! perfect!! I missed it.

File: zkevm-circuits/src/evm_circuit.rs
Comment (CPerezz) [reply zkevm-circuits/src/evm_circuit.rs:35]: Nonono!! It's not dangerous. And it's meant to be used as `&impl LookupTable` for example.

But using `dynamic dispatch` brings in dyn-safety concerns when using generics. It's fine as this is the correct solution. But I just raise the flag as dyn-safety rules are pretty restrictive. So we will need to play with references and explicit lifetimes more.

All I mean is that `LookupTable` trait will not be able to have certain methods or logic. And we need to count on it from now on :)

See: https://doc.rust-lang.org/nightly/reference/items/traits.html#object-safety

See also a really interesting article (in case you are curious and like Rust) where `dyn*` is discussed and the caveats tha `& dyn` currently has: https://smallcultfollowing.com/babysteps/blog/2022/03/29/dyn-can-we-make-dyn-sized/

File: zkevm-circuits/src/evm_circuit.rs
Comment (CPerezz) [reply zkevm-circuits/src/evm_circuit.rs]: Yes, I mainly see two solutions:

1. You have a private internal fn which has the bool field and you expose two public fns which pass the `true` or `false` and mark one with `#[cfg(test)]` (More work IMO).

2. You have two functions with the exact same name and signature, and internally you define the logic taking into account the bool implications. You then mark one fn (the one with true IIUC) with `#[cfg(test)]` and you mark the other with: `#[cfg(not(test))]`

So that the compiler treats them as mutably exclusive. 

Whatever you prefer will be better than having this bool in the API :)

File: zkevm-circuits/src/evm_circuit/util.rs
Comment (CPerezz) [reply zkevm-circuits/src/evm_circuit/util.rs:120]: Ohhh I see! 

Thanks for the explanation!!! :)

File: zkevm-circuits/src/evm_circuit/util.rs
Comment (CPerezz) [reply zkevm-circuits/src/evm_circuit/util.rs:279]: Makes sense. 

If the number of accesses is not big it's fine to use the `BTreeMap`.
If we call this a lot of times, then we need to take into account that `BTreeMap` storage is slower. So depending on the frequency which with this code is called, maybe we can try to think about better ways in the future :smile:

File: zkevm-circuits/src/evm_circuit/execution.rs
Comment (han0110) [reply zkevm-circuits/src/evm_circuit/execution.rs:239]: Sounds good to me! Just opened issue #464.

File: zkevm-circuits/src/evm_circuit.rs
Comment (Brechtpd) [reply zkevm-circuits/src/evm_circuit.rs]: I picked option 1 because I think it's probably useful to allow both variants to be used when testing (if they behave differently it could potentially hide some problems in the non-test version that may need specific testing).

Comment (CPerezz) [reply zkevm-circuits/src/evm_circuit.rs]: Hmm makes sense!!! LGTM then!!!

**Review by han0110 — APPROVED** [review]
LGTM! Only got some nitpicks

File: zkevm-circuits/src/evm_circuit/util/constraint_builder.rs
Comment (Brechtpd) [reply zkevm-circuits/src/evm_circuit/util/constraint_builder.rs]: Yep not needed anymore, but yeah I thought it could still be useful somehow but, at least how things currently work, I can't find any use cases so removed it.

And great idea about the duplicated lookups, will do!

File: zkevm-circuits/src/evm_circuit/execution.rs
Comment (Brechtpd) [reply zkevm-circuits/src/evm_circuit/execution.rs:722]: I think it's required because I do believe it gets accessed in the witness generation for the stored expressions. The only reason it seems to work is because `MAX_STEP_HEIGHT` is currently big enough so that most of the times it also included some more rows than just needed for the two execution states that are actually being processed. I changed the height to perfectly match the height actually required for the current and next step and I got an out of bounds error (matching the height perfectly I feel doesn't make sense because it makes the logic a bit more complicated for really no reason as the region is small anyway).

File: zkevm-circuits/src/evm_circuit/param.rs
Comment (Brechtpd) [reply zkevm-circuits/src/evm_circuit/param.rs:4]: Pretty much yeah. Because most opcodes currently still take up at least 2 rows I would say it's very likely that making the circuit even wider (more columns and/or extra lookups per row) so most cheap (cheap as in gas cost) opcodes take up 1 row will very likely be better.

**Review by CPerezz — APPROVED** [review]
LGTM modulo @han0110 's review comments addressed! :smile:

**Review by han0110 — APPROVED** [review]
All my comments are addressed! Let's merge this after conflicts are resolved.

**Comment by CPerezz** [comment]
@Brechtpd , when you can, just resolve the conflicts and merge! :smile:

**Comment by CPerezz** [comment]
Aaaaah!! True! Will release  a new tag and ping you in here!

**Comment by CPerezz** [comment]
@Brechtpd here you go: https://github.com/appliedzkp/halo2/releases/tag/v2022_05_06

I'll make a PR to update the tag and you can rebase on the top once merged and leave it ready for merge.

**Comment by CPerezz** [comment]
As reported by @ed255 `test evm_circuit::execution::call::test::call_gadget_recursive` is the test that has taken all that time. 

Any particular reason you consider @Brechtpd ?

**Comment by Brechtpd** [comment]
The reason for the terrible performance addressed here: https://github.com/appliedzkp/halo2/pull/71. The evm circuit gates reference a lot of cells and this is even more true in this PR, so finding cells in a Vec is slow. Only really important for circuits that use a lot of rows, which currently is only `call_gadget_recursive`. What also didn't help for this is that cells are generally assigned twice now. This isn't really important for the tests after the fix but probably still worth optimizing for the actual prover times, but will leave that to a later PR that can focus on witness assignment optimizations.

I applied @han0110's `call_gadget_recursive` test change to this PR: https://github.com/appliedzkp/zkevm-circuits/commit/86e7dfaa7e4bf5f2b0eb06e329b44a015ac30568. This already makes the test pass in a reasonable time. But after applying [halo2 mock prover improvement above](https://github.com/appliedzkp/halo2/pull/71) the test time of that test is reduced by another 10x.

Compared to the current master (both with Han's patch applied and https://github.com/appliedzkp/halo2/pull/71),  `call_gadget_recursive` will actually be more than 2x faster after this PR (the reduced number of lookups helps both the MockProver and the actual prover). In general though the tests will be slower after this PR even with https://github.com/appliedzkp/halo2/pull/71 mainly because of the more expensive circuit building to optimize the constraints but the difference isn't that big (around 1.5x slower). But if a test uses a lot of rows than the test will run faster.

**Comment by CPerezz** [comment]
That sounds amazing!!!

Nice job @Brechtpd !!!
I'll try to also apply [`cargo nextest`](https://nexte.st) and [`rust-cache`](https://github.com/marketplace/actions/rust-cache) to see if we can squeeze some extra time savings.

**Comment by CPerezz** [comment]
Once https://github.com/appliedzkp/zkevm-circuits/pull/503 gets merged, we're good to go! (I hope.. :sweat_smile: ahahah

**Comment by CPerezz** [comment]
@Brechtpd , the new tag is at the head of `main`. 

Can you rebase and let's see what happens?

**Comment by Brechtpd** [comment]
Now only ~2 minutes slower to run all tests compared to the master, not bad I think. :)

**Pull request merged**
